tests/components/recorder/test_migration_run_time_migrations_remember.py

```diff
@@ -0,0 +1,163 @@
+"""Test run time migrations are remembered in the migration_changes table."""
+
+import importlib
+from pathlib import Path
+import sys
+from unittest.mock import patch
+
+import pytest
+from sqlalchemy import create_engine
+from sqlalchemy.orm import Session
+
+from homeassistant.components import recorder
+from homeassistant.components.recorder import core, migration, statistics
+from homeassistant.components.recorder.queries import get_migration_changes
+from homeassistant.components.recorder.tasks import StatesContextIDMigrationTask
+from homeassistant.components.recorder.util import (
+    execute_stmt_lambda_element,
+    session_scope,
+)
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import HomeAssistant
+
+from .common import async_recorder_block_till_done, async_wait_recording_done
+
+from tests.common import async_test_home_assistant
+from tests.typing import RecorderInstanceGenerator
+
+CREATE_ENGINE_TARGET = "homeassistant.components.recorder.core.create_engine"
+SCHEMA_MODULE = "tests.components.recorder.db_schema_32"
+
+
+async def _async_wait_migration_done(hass: HomeAssistant) -> None:
+    """Wait for the migration to be done."""
+    await recorder.get_instance(hass).async_block_till_done()
+    await async_recorder_block_till_done(hass)
+
+
+def _get_migration_id(hass: HomeAssistant) -> dict[str, int]:
+    with session_scope(hass=hass, read_only=True) as session:
+        return dict(execute_stmt_lambda_element(session, get_migration_changes()))
+
+
+def _create_engine_test(*args, **kwargs):
+    """Test version of create_engine that initializes with old schema.
+
+    This simulates an existing db with the old schema.
+    """
+    importlib.import_module(SCHEMA_MODULE)
+    old_db_schema = sys.modules[SCHEMA_MODULE]
+    engine = create_engine(*args, **kwargs)
+    old_db_schema.Base.metadata.create_all(engine)
+    with Session(engine) as session:
+        session.add(
+            recorder.db_schema.StatisticsRuns(start=statistics.get_start_time())
+        )
+        session.add(
+            recorder.db_schema.SchemaChanges(
+                schema_version=old_db_schema.SCHEMA_VERSION
+            )
+        )
+        session.commit()
+    return engine
+
+
+@pytest.mark.parametrize("enable_migrate_context_ids", [True])
```

Comment:
This had to be its own file because we auto use schema 32 in the other migration tests and we want restart HA again with the current schema version in the middle of the test to verify the migration behavior